modtools/fire-rate.lua

```diff
@@ -0,0 +1,511 @@
+  if ( dfhack.world.isFortressMode() ) then
+    current_game_mode = df.game_mode.DWARF
+  elseif ( dfhack.world.isAdventureMode() ) then
+    current_game_mode = df.game_mode.ADVENTURE
+  elseif ( dfhack.world.isArena() ) then
+    current_game_mode = df.global.gamemode
+  end
+
+  local fire_rate = get_weapon_fire_rate(weapon_token, material_token,
+                                         current_game_mode)
+  local speed_cap = get_weapon_fire_rate_cap(weapon_token, material_token,
+                                             current_game_mode)
+  local skill_factor = get_weapon_skill_factor(weapon_token, material_token,
+                                               current_game_mode)
+
+  -- Filter method based on current mode
+  -- (TODO: Except not actually because only one is available)
+  if ( get_config_skill_mode() == "basic" or true ) then
+    local uncapped_rate = math.floor(
+                            fire_rate - (unit_skill_level * skill_factor) + 0.5)
+
+    return math.max(uncapped_rate, speed_cap)
+  end
+end
+
+-- First rolls a skill check to determine how long it'll take for the unit to
+-- recover, then updates the unit's recovery time accordingly
+function do_shot_cooldown(unit, weapon_token, material_token)
+  local fire_rate = get_fire_rate_for_unit(unit, weapon_token, material_token)
+
+  apply_shot_cooldown(unit, fire_rate)
+end
+
+-- Causes the unit to have to wait the given length of time (in ticks) before
+-- they can act again
+-- This will account for the hard-coded delay present in adventure actions
+function apply_shot_cooldown(unit, recovery_time)
+  local time_to_use = recovery_time
+
+  -- Account for the hard-coded recovery time added for adventure mode actions
+  -- performed by the player (AI aren't subject to the delay)
+  if ( df.global.gamemode == df.game_mode.ADVENTURE
+            and unit.id == get_adventurer_unit().id ) then
+    -- The hardcoded delay is still active even in tactical mode, so no need to
+    -- try and account for it
+    time_to_use = math.max(recovery_time - 9, 0)
+  end
+
+  unit.counters.think_counter = time_to_use
+end
+
+---------------------------------------------------------------------
+function get_weapon_skill_ranged(weapon_token)
+  if ( weapon_token == throw_key ) then
+    return df.job_skill.THROW
+  end
+
+  local item_def = get_item_def_by_token(weapon_token)
+  -- Only tools and weapons have a `skill_ranged` value
+  if ( df.itemdef_toolst:is_instance(item_def)
+            or df.itemdef_weaponst:is_instance(item_def) ) then
+    return item_def.skill_ranged
+  else
+    return df.job_skill.NONE -- NONE, or default to throw?
+  end
+end
+
+-- Returns the item def of the provided item
+-- (will technically clash if two items in different categories use the same
+-- token, but I don't know if that's allowed anyway)
+function get_item_def_by_token(item_token)
+  local subtype_token = utils.split_string(item_token, ":")[2]
```

Comment:
consider: `local subtype_token = item_token:split(':')[2]`

Reply:
I was originally going to object to that change on the grounds that it unnecessarily obfuscates where the split function was coming from, but then I realised it also doesn't come from `utils` either. Where *does* split come from? And what's the point of `utils` having the `split_string` function?

Reply:
sorry, I thought I replied to this earlier. the `string:split()` function is defined in `dfhack.lua` so it's available to all scripts:
https://github.com/DFHack/dfhack/blob/develop/library/lua/dfhack.lua#L413

the function used to be defined in `utils.lua`. The definition was recently (within the last year) moved to `dfhack.lua` so it was accessible as a `string` function along with the other related string functions that `dfhack.lua` provides.